File: tests/model_loader.py
from buzz.model_loader import TranscriptionModel, ModelDownloader


def get_model_path(transcription_model: TranscriptionModel) -> str:
    path = transcription_model.get_local_model_path()
    if path is not None:
        return path

    model_loader = ModelDownloader(model=transcription_model)
    model_path = ""

    def on_load_model(path: str):
        nonlocal model_path
        model_path = path

    model_loader.signals.finished.connect(on_load_model)
    model_loader.run()
    return model_path

File: Buzz.swift/Buzz/ModelLoader.swift
//
//  ModelLoader.swift
//  Buzz
//
//  Created by Chidi Williams on 08/02/2023.
//

import Foundation

class ModelLoader {
    static func isAvailable(model: WhisperModel) -> Bool {
        do {
            let modelPath = try getModelPath(model: model)
            return FileManager.default.fileExists(atPath: modelPath.path(percentEncoded: false))
        } catch {
            print(error)
            return false
        }
    }
    
    static func getModelPath(model: WhisperModel) throws -> URL  {
        return try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
            .appending(path: "Buzz")
            .appending(path: "Models")
            .appendingPathComponent("ggml-whisper-\(model.id).bin")
    }
    
    static func getModelByteSize(model: WhisperModel) -> Int64 {
        switch model {
        case .tiny, .tiny_en:
            return 77_700_000
        case .base, .base_en:
            return 148_000_000
        case .small, .small_en:
            return 488_000_000
        case .medium, .medium_en:
            return 1_530_000_000
        case .large:
            return 3_090_000_000
        }
    }
    
    static func getModelDownloadURL(model: WhisperModel) -> URL? {
        return URL(string: "https://huggingface.co/datasets/ggerganov/whisper.cpp/resolve/main/ggml-\(model.id).bin")
    }
}
